[MRG] Stage 1 of updates to Dataset encoding state and writing

scaramallion opened the pull request:
#### Describe the changes

##### Stage 1 (this PR)
###### Encoding
* Make `Dataset.read_implicit_vr` and `Dataset.read_little_endian` read-only, deprecate for removal in v4.0
* Add `Dataset.original_encoding`: the encoding in the original file, read-only (replacement for `read_implicit_vr`, `read_little_endian`
* Deprecate `Dataset.is_implicit_VR` and `Dataset.is_little_endian` for removal in v4.0

###### Character Set
* Add `Dataset.original_character_set`: character set from original file, read-only   
* Deprecated `Dataset.read_encoding` for removal in v4.0

###### Other
* Be a bit more thorough in ensuring the file meta is conformant
* Clean up the previously existing warnings not being caught by pytest

##### Stage 2 (to be done in a separate PR)
Change `Dataset.save_as()` to:
```python
def save_as(
    self, 
    filename, 
    __write_like_original = True,  # to be deprecated 
    implicit_vr: bool | None = None, 
    little_endian: bool = True,
    enforce_file_format: bool = False,
    **kwargs,   # will handle backwards compatibility with `write_like_original`
):
```
Change `dcmwrite()` in a similar manner. 

The logic here is threefold:
 1. Encoding options should either be specified through transfer syntax or if not available through explicit arguments to `save_as()` and `dcmwrite()`. 
 2. `save_as()` creates a copy of the dataset (as opposed to a non-existent `Dataset.save()` function), so it shouldn't be modifying the original as part of the saving process. 
 3. `write_like_original` doesn't actually do anything except act as a flag for whether DICOM File Format is enforced or not (and a niche command set writing override that can be done instead with `write_dataset()`)

`Dataset.is_implicit_VR` and `Dataset.is_little_endian` should will be eventually removed. `read_implicit_VR` and `read_little_endian` will also be removed - replaced with the read-only `Dataset.original_encoding`. 

#### Tasks
- [x] Unit tests added that reproduce the issue or prove feature is working
- [x] Fix or feature added
- [x] Code typed and mypy shows no errors
- [x] Documentation updated (if relevant)
- [x] Unit tests passing and overall coverage the same or better

scaramallion commented on src/pydicom/util/codify.py:
```diff
@@ -413,8 +413,8 @@ def code_file_from_dataset(
         lines.append("ds.file_meta = file_meta")
-    lines.append("ds.is_implicit_VR = " + str(ds.is_implicit_VR))
-    lines.append("ds.is_little_endian = " + str(ds.is_little_endian))
+    lines.append("ds._is_implicit_VR = " + str(ds.is_implicit_VR))
+    lines.append("ds._is_little_endian = " + str(ds.is_little_endian))
```
Won't be necessary after Stage 2

scaramallion commented on tests/test_dataelem.py:
```diff
@@ -472,7 +472,7 @@ def check_empty_text_element(value):
         config.use_none_as_empty_text_VR_value = use_none
         ds = Dataset()
-        ds.is_little_endian = True
+        ds._is_little_endian = True
```
A lot of these explicit settings of `_is_little_endian` and `_is_implicit_VR` in the tests get rolled in the `save_as()`  changes in Stage 2

scaramallion commented on src/pydicom/pixel_data_handlers/util.py:
```diff
@@ -1267,7 +1267,7 @@ def pixel_dtype(ds: "Dataset", as_float: bool = False) -> "np.dtype":
 
     if ds.is_little_endian is None:
-        ds.is_little_endian = ds.file_meta.TransferSyntaxUID.is_little_endian
+        ds._is_little_endian = ds.file_meta.TransferSyntaxUID.is_little_endian
```
WIll refactor this in Stage 2 to avoid prioritise transfer syntax and fall back to `is_little_endian`

scaramallion commented on tests/test_valuerep.py:
```diff
@@ -917,7 +917,8 @@ def test_float_value(self):
         with BytesIO(bin_elem) as bio:
             ds = read_dataset(bio, True, True)
-        assert isinstance(ds.Exposure, ISfloat)
+        with pytest.warns(UserWarning, match="Invalid value for VR IS"):
```
Unrelated, cleaned up the warnings not getting caught by pytest

mrbean-bremen reviewed (approved):
Looks good to me, thanks! I think this stuff was always quite confusing for pydicom users (it certainly was for me), and it is a good thing to clean this up.

darcymason commented:
I support these changes for sure, but have a couple of thoughts around details...:

> `Dataset.is_implicit_VR` and `Dataset.is_little_endian` should then be demoted to only be the encoding used in the original dataset or `None` if no original dataset. As these will become read-only, `read_implicit_VR` and `read_little_endian` will no longer be needed. So to make reviewing a bit easier I've split these changes out first.

I'm not sure about the naming proposed.  Largely because of the history, perhaps the one that should be kept are the `read_...` names, if that is what they will actually represent, and that would help avoid confusion with the deprecated `is_...` forms.

The other thought is that the new `save_as` (etc.) proposed does not fully deal with setting transfer syntax.  Looking again at the UID class bool functions `is_little_endian`, `is_implicit_VR`, and `is_compressed`, they really are pretty straightforward (one implicit, otherwise explicit; one big endian else little-endian).  I think the proposed function parameters is better than what we have now, but I wonder if there is even more elegant ways, especially if more compressed forms are eventually available.  We will still have logic to error after the fact, if e.g. someone passes something incompatible with the file_meta t-syntax.  I'm torn though because I feel like proposing something more is at risk of unnecessarily complicating things, like making users import long-named UIDs for simple cases.  I'd be happy to entertain any thoughts on this.

That's for step 2, but perhaps if discussed here first it might set a slightly different direction.

scaramallion commented:
> I'm not sure about the naming proposed. Largely because of the history, perhaps the one that should be kept are the read_... names, if that is what they will actually represent, and that would help avoid confusion with the deprecated is_... forms.

Yeah, I've been thinking about it while working on stage 2 and I agree. I think I'll make `is_little_endian` and `is_implicit_VR` reflect the encoding as currently set in the dataset (but still read-only), so they'd be:
```python
>>> ds = dcmread(...) # read an implicit little
>>> ds.read_implicit_vr, ds.read_little_endian
(True, True)
>>> ds.is_implicit_VR, ds.is_little_endian
True, True
>>> ds.file_meta.TransferSyntaxUID = ExplicitVRBigEndian
>>> ds.is_implicit_VR, ds.is_little_endian
False, True
```

I was hoping I could avoid having to add `FileMetaDataset.parent` but I think I may have to

scaramallion commented:
> I'm not sure about the naming proposed. Largely because of the history, perhaps the one that should be kept are the read_... names, if that is what they will actually represent, and that would help avoid confusion with the deprecated is_... forms.

Yep, agreed

scaramallion commented:
OK, updated. Summary:

* `is_little_endian` and `is_implicit_VR` deprecated for removal, no replacement
* `read_little_endian` and `read_implicit_VR` deprecated for removal, replaced by `original_encoding` (read-only property)
* `read_encoding` deprecated for removal, replaced by `original_character_set` (read-only property)

darcymason commented:
> OK, updated. Summary:
> 
> * `is_little_endian` and `is_implicit_VR` deprecated for removal, no replacement
> * `read_little_endian` and `read_implicit_VR` deprecated for removal, replaced by `original_encoding` (read-only property)
> * `read_encoding` deprecated for removal, replaced by `original_character_set` (read-only property)

I like the names -- I had also thought about changing the `read_encoding` to something like `original_encoding`, it conveys the meaning better.

darcymason commented on src/pydicom/dataset.py:
```diff
@@ -911,14 +895,63 @@ def __getattr__(self, name: str) -> Any:
+            ),
+            DeprecationWarning,
+        )
+
```
Do we want to continue the [future behavior idea](https://pydicom.github.io/pydicom/stable/old/best_practices.html#future-proofing-your-code)? (I think it can be valuable to some people).  If so, then these should throw an error if 'future' is on.

scaramallion replied:
It's up to you, I'm not fussed either way. I might add it in stage 2 though, just so I can progress forwards on this one?

darcymason replied:
Sure, stage 2 is fine
